client/transaction-pool/src/lib.rs

```diff
@@ -727,6 +774,220 @@ where
+}
+
+struct EnactmentHelper<Block, PoolApi>
+where
+	Block: BlockT,
+	PoolApi: graph::ChainApi<Block = Block>,
+{
+	best_block: Arc<Mutex<Option<Block::Hash>>>,
+	finalized_block: Arc<Mutex<Option<Block::Hash>>>,
+	metadata_cache: Arc<Mutex<HeaderMetadataCache<Block>>>,
+	api: Arc<PoolApi>,
+}
+
+impl<Block, PoolApi> EnactmentPolicy<Block> for EnactmentHelper<Block, PoolApi>
+where
+	Block: BlockT,
+	PoolApi: 'static + graph::ChainApi<Block = Block>,
+{
+	fn should_handle_enactment(
+		&self,
+		event: &ChainEvent<Block>,
+	) -> (bool, Arc<Vec<Block::Hash>>, Arc<Vec<Block::Hash>>) {
+		match event {
+			ChainEvent::NewBestBlock { hash, tree_route } => {
+				println!("--------------------------------------------------------------------------------");
+				println!("process_new_best_block_event");
+				let (enacted, retracted) = if let Some(tree_route) = tree_route {
+					(
+						tree_route.enacted().iter().map(|block| block.hash).collect(),
+						tree_route.retracted().iter().map(|block| block.hash).collect(),
+					)
+				} else {
+					(vec![], vec![])
+				};
+
+				let result = self.resolve(*hash, &enacted, &retracted, false);
+
+				(result, Arc::from(enacted), Arc::from(retracted))
+			},
+			ChainEvent::Finalized { hash, tree_route } => {
+				println!("--------------------------------------------------------------------------------");
+				println!("process_finalized_event");
+				let tree_route = Arc::new(tree_route.to_vec()); //todo: cleanup
+				let result = self.resolve(*hash, &tree_route, &vec![], true);
+				(result, tree_route, Arc::from(vec![]))
+			},
+		}
+	}
+}
+
+impl<Block, PoolApi> EnactmentHelper<Block, PoolApi>
+where
+	Block: BlockT,
+	PoolApi: 'static + graph::ChainApi<Block = Block>,
+{
+	fn new(api: Arc<PoolApi>) -> Self {
+		EnactmentHelper {
+			best_block: Arc::new(Mutex::new(None)),
+			finalized_block: Arc::new(Mutex::new(None)),
+			metadata_cache: Arc::new(Mutex::new(HeaderMetadataCache::new(16))),
+			api,
+		}
+	}
+
+	fn resolve(
+		&self,
+		hash: Block::Hash,
+		enacted: &Vec<Block::Hash>,
+		retracted: &Vec<Block::Hash>,
+		finalized: bool,
+	) -> bool {
+		println!(
+			"--------------------------------------------------------------------------------"
+		);
+		println!("xxx handle_enactment hash:{hash:?} finalized:{finalized:?} retracted: {retracted:?} enacted: {enacted:?}");
+
+		//update the metadata_cache with all incoming data
+		{
+			let cache = self.metadata_cache.lock();
+			enacted.iter().chain(retracted.iter()).chain(std::iter::once(&hash)).for_each(
+				|&hash| {
+					let id = BlockId::hash(hash);
+					if let Ok(Some(header)) = self.api.block_header(&id) {
+						println!("into cache: {:?} {:?}", hash, header);
+						cache.insert_header_metadata(hash, CachedHeaderMetadata::from(&header));
+					}
+				},
+			);
+		}
+
+		//compute the tree_route from recently finalized block to provided hash
+		//check if recently finalized block is on retracted path
```

Comment:
Why don't you just use the `tree_route` function?

Reply:
Done.